Language/Haskell/GhcMod/Monad/Types.hs

```diff
@@ -129,10 +138,11 @@ data GhcModState = GhcModState {
       gmGhcSession   :: !(Maybe GmGhcSession)
     , gmComponents   :: !(Map ChComponentName (GmComponent GMCResolved (Set ModulePath)))
     , gmCompilerMode :: !CompilerMode
+    , gmMMappedFiles :: Map TargetId (StringBuffer, UTCTime)
```

Comment:
I will answer the main bulk of your comment probably tomorrow, since I'm pretty sleepy now, but a couple more facts that you may not be aware of:
 - GHC API requires a file to exist when loading targets. Its contents can be supplied via argument, but file itself *has* to be there, as far as I can tell, although it can be empty, or have arbitrary contents.
 - GHC API forbids CPP preprocessing on in-memory files (more like, it just can't handle it on its own, actually). It's generally possible to query for types etc, but when running a check, it just bails.

So no, building a could IDE based just on this mechanic is not feasible. It's handy when running a type query on a modified-but-not-yet-saved file, but that's pretty much the limit of its usefulness.

> But that's exactly what you have to do to get ghc to accept your remapped module, is it not? .. I'm so confused.

For now, I just load mappings that are explicitly passed as targets, not *all* mapped files (i.e. check file.hs would only load a mapping for file.hs, if it's specified -- for my use-case that's actually quite enough). But you have a point there -- if there is no need to load a file, there is no reason to define a mapping for it, so I should probably reevaluate that.

> But only if the file actually exists IIRC from the docs.
As I said, GHC API requires file to actually exist to supply a source for it, but even if corresponding file doesn't exist, if a target "looks" like filename, i.e. has `.hs`/`.lhs` suffix, then `guessTarget` interprets it as filename regardless. Also if adding `.hs` or `.lhs` to argument string yields an existing filename, it will be interpreted as filename. If everything else fails, only then it would be interpreted as a module name (that is, if it doesn't contain characters forbidden in module name, otherwise it just throws)